### PR TITLE
[16.0][MIG] stock_no_negative

### PR DESCRIPTION
**Pull request by vvanneri-ia**
Please Review

### CODE REVIEW

**Comment by vvanneri-ia** [comment]
@rousseldenis @pedrobaeza  Please Check it

File: stock_no_negative/readme/HISTORY.rst [hunk 1]
```diff
@@ -0,0 +1,15 @@
+14.0.1.0.0 (2020-12-14)
```
Comment (rousseldenis) [inline stock_no_negative/readme/HISTORY.rst:11]: Could you update changelog ?

File: stock_no_negative/models/stock_quant.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            if (
+                float_compare(quant.quantity, 0, precision_digits=p) == -1
+                and quant.product_id.type == "product"
+                and quant.location_id.usage in ["internal", "transit"]
+                and disallowed_by_product
+                and disallowed_by_location
+            ):
+                msg_add = ""
+                if quant.lot_id:
+                    msg_add = _(" lot '%s'") % quant.lot_id.name_get()[0][1]
```
Comment (rousseldenis) [inline stock_no_negative/models/stock_quant.py]: Could you use format() instead ?

File: stock_no_negative/models/stock_quant.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                raise ValidationError(
+                    _(
+                        "You cannot validate this stock operation because the "
+                        "stock level of the product '%(name)s'%(name_lot)s would "
+                        "become negative "
+                        "(%(q_quantity)s) on the stock location '%(complete_name)s' "
+                        "and negative stock is "
+                        "not allowed for this product and/or location."
+                    )
+                    % {
```
Comment (rousseldenis) [inline stock_no_negative/models/stock_quant.py]: Same as above

File: stock_no_negative/models/stock_quant.py (hunk 2 above)
Comment (rousseldenis) [reply stock_no_negative/models/stock_quant.py]: Could you also add a little test for quants with lots ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

**Review by marylla — APPROVED** [review]
I did a functional review. 

* I created an outgoing delivery for a product with zero stock but with "allow negative stock = True". I could validate the picking and the stock on hand was negative.

* In a second outgoing delivery the validation was blocked for product with zero stock but with "allow negative stock = False".

* In a third outgoing delivery the validation was **not** blocked for a product with zero stock and with "allow negative stock = False" but with "allow negative stock = True" for the product category.

* In a 4th outgoing delivery the delivery was **not** blocked for a a product with zero stock and with "allow negative stock = False" but with "allow negative stock = True" for its source location.

So it works like in previous Odoo versions. :+1:

File: stock_no_negative/readme/HISTORY.rst (hunk 1 above)
Comment (vvanneri-ia) [reply stock_no_negative/readme/HISTORY.rst:11]: for 15.0?

Comment (rousseldenis) [reply stock_no_negative/readme/HISTORY.rst:11]: At lease for this one

File: stock_no_negative/models/stock_quant.py (hunk 3 above)
Comment (vvanneri-ia) [reply stock_no_negative/models/stock_quant.py]: do i need to change the pot files?

**Comment by florenciafrigieri2** [comment]
@vvanneri-ia Hi! Are you going to continue with this PR? Thanks in advance!

**Comment by vvanneri-ia** [comment]
@florenciafrigieri2 Yes will update asap

File: stock_no_negative/models/stock_quant.py (hunk 3 above)
Comment (rousseldenis) [reply stock_no_negative/models/stock_quant.py]: Yes

**Comment by rousseldenis** [comment]
@vvanneri-ia Could you attend comments?

**Review by alexis-via — APPROVED** [review]
code review (code is unchanged !) and tests

**Comment by vvanneri-ia** [comment]
@rousseldenis Could you have look?

**Comment by rousseldenis** [comment]
@vvanneri-ia Could you fix tests ?

**Comment by Murtaza-OSI** [comment]
@vvanneri-ia Can you fix test case.

**Comment by Zar-SW** [comment]
Any updates regarding this PR?

**Comment by rousseldenis** [comment]
@vvanneri-ia

**Comment by desdelinux** [comment]
ping @vvanneri-ia

**Comment by vvanneri-ia** [comment]
Sure

**Comment by vvanneri-ia** [comment]
@rousseldenis Please check the error, which i felt is odoo error ,
image

**Comment by alexis-via** [comment]
Tests are running fine on my laptop. So I think a rebase would re-trigger tests and make it green.

**Review by rousseldenis — APPROVED** [review]
Code review

**Pull request merged**